presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -76,42 +85,95 @@ public PageProcessor(Optional<PageFilter> filter, List<? extends PageProjection>
 
     public PageProcessorOutput process(ConnectorSession session, DriverYieldSignal yieldSignal, Page page)
     {
-        // limit the scope of the dictionary ids to just one page
-        dictionarySourceIdFunction.reset();
+        AggregatedMemoryContext memoryContext = newSimpleAggregatedMemoryContext();
+        return new PageProcessorOutput(memoryContext::getBytes, process(session, yieldSignal, memoryContext, page));
+    }
+
+    public Iterator<Optional<Page>> process(ConnectorSession session, DriverYieldSignal yieldSignal, AggregatedMemoryContext memoryContext, Page page)
+    {
+        WorkProcessor<Page> processor = process(session, yieldSignal, memoryContext, WorkProcessor.create(new SingletonPage(memoryContext, page)));
+        return processor.yieldingIterator();
+    }
+
+    public WorkProcessor<Page> process(ConnectorSession session, DriverYieldSignal yieldSignal, AggregatedMemoryContext memoryContext, WorkProcessor<Page> pages)
+    {
+        return pages.flatTransform(new PageFilterAndProject(session, yieldSignal, memoryContext));
+    }
+
+    private class SingletonPage
+            implements WorkProcessor.Process<Page>
+    {
+        private final LocalMemoryContext memoryContext;
+        private Page page;
 
-        if (page.getPositionCount() == 0) {
-            return EMPTY_PAGE_PROCESSOR_OUTPUT;
+        private SingletonPage(AggregatedMemoryContext memoryContext, Page page)
+        {
+            this.memoryContext = memoryContext.newLocalMemoryContext();
+            this.page = page;
+            this.memoryContext.setBytes(calculateRetainedSizeWithoutLoading(page));
         }
 
-        if (filter.isPresent()) {
-            SelectedPositions selectedPositions = filter.get().filter(session, filter.get().getInputChannels().getInputChannels(page));
-            if (selectedPositions.isEmpty()) {
-                return EMPTY_PAGE_PROCESSOR_OUTPUT;
+        @Override
+        public ProcessorState<Page> process()
+        {
+            if (page == null) {
+                return finished();
             }
 
-            if (projections.isEmpty()) {
-                return new PageProcessorOutput(() -> calculateRetainedSizeWithoutLoading(page), singletonIterator(Optional.of(new Page(selectedPositions.size()))));
+            Page result = page;
+            page = null;
+            memoryContext.close();
+            return ProcessorState.ofResult(result);
+        }
+    }
+
+    private class PageFilterAndProject
```

Comment:
If the plan is to use this class in other contexts, I think you will need to be careful about the fact this is not a static class and it updating data that is shared across all executions (auto tuning of the work size).